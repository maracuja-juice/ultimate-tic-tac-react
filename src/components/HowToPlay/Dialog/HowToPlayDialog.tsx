import * as React from 'react';
import { useState } from 'react';
import {
  Dialog,
  DialogActions,
  DialogButton,
  DialogContent,
  DialogTitle,
} from '@rmwc/dialog';
import { HowToPlayContent } from '../Content/HowToPlayContent';
import icons from '../../../icons/icons';
import { steps } from '../HowToPlaySteps';
import styles from './HowToPlayDialog.module.css';
import { DialogOnCloseEventT } from '@rmwc/dialog/dist/dialog';

export interface HowToPlayDialogProps {
  onClose: (evt: DialogOnCloseEventT) => void;
}

export function HowToPlayDialog(props: HowToPlayDialogProps) {
  const [stepNumber, setStepNumber] = useState(0);
  const step = steps[stepNumber];

  return (
    <Dialog className={styles.dialog} open={true} onClose={props.onClose}>
      <DialogTitle>How to play</DialogTitle>
      <DialogContent>
        <HowToPlayContent
          text={step.text}
          boardStates={step.states}
          moves={step.moves}
        />
      </DialogContent>
      <DialogActions>
        <DialogButton action="close">Cancel</DialogButton>
        <DialogButton
          className={styles.push}
          icon={{ icon: icons.arrowLeft, 'aria-hidden': true }}
          disabled={stepNumber === 0}
          onClick={() => {
            setStepNumber(stepNumber - 1);
          }}
        >
          Previous
        </DialogButton>
        <DialogButton
          trailingIcon={{ icon: icons.arrowRight, 'aria-hidden': true }}
          disabled={stepNumber === steps.length - 1}
          onClick={() => {
            setStepNumber(stepNumber + 1);
          }}
          isDefaultAction
        >
          Next
        </DialogButton>
      </DialogActions>
    </Dialog>
  );
}
