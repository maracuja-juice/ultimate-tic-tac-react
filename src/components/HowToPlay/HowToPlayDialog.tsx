import * as React from 'react';
import { Button } from '@rmwc/button';
import {
  Dialog,
  DialogActions,
  DialogButton,
  DialogContent,
  DialogTitle,
} from '@rmwc/dialog';
import { BigBoard } from '../Board/BigBoard/BigBoard';
import { Player, TileValue } from '../../state/AppState';
import { SimpleTicTacToeGame } from '../../util';

export function HowToPlayDialog() {
  const [open, setOpen] = React.useState(false);

  const simpleState = [
    TileValue.Cross,
    TileValue.Cross,
    TileValue.Cross,
    TileValue.Circle,
    TileValue.Circle,
    TileValue.Empty,
    TileValue.Empty,
    TileValue.Empty,
    TileValue.Empty,
  ];
  const board = new SimpleTicTacToeGame(simpleState).getBoard();

  return (
    <>
      <Dialog
        open={open}
        onClose={(evt) => {
          console.log(evt.detail.action);
          setOpen(false);
        }}
        onClosed={(evt) => console.log(evt.detail.action)}
      >
        <DialogTitle>Dialog Title</DialogTitle>
        <DialogContent>
          <BigBoard
            currentPlayer={Player.Circle}
            board={board}
            activeBoards={[]}
          />
        </DialogContent>
        <DialogActions>
          <DialogButton action="close">Cancel</DialogButton>
          <DialogButton action="accept" isDefaultAction>
            Sweet!
          </DialogButton>
        </DialogActions>
      </Dialog>

      <Button raised onClick={() => setOpen(true)}>
        Open standard Dialog
      </Button>
    </>
  );
}
