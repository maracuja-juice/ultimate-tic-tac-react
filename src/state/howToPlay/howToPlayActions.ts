export const HOW_TO_PLAY_OPEN = 'howToPlayReducer/how-to-play-open';
export const HOW_TO_PLAY_CLOSE = 'howToPlayReducer/how-to-play-close';
export const HOW_TO_PLAY_STEP_FORWARD =
  'howToPlayReducer/how-to-play-step-forward';
export const HOW_TO_PLAY_STEP_BACKWARD =
  'howToPlayReducer/how-to-play-step-backward';
export const HOW_TO_PLAY_STATE_FORWARD =
  'howToPlayReducer/how-to-play-state-forward';

export const howToPlayOpen = () => ({
  type: HOW_TO_PLAY_OPEN,
});

export const howToPlayClose = () => ({
  type: HOW_TO_PLAY_CLOSE,
});

export const howToPlayStepForward = () => ({
  type: HOW_TO_PLAY_STEP_FORWARD,
});

export const howToPlayStepBackward = () => ({
  type: HOW_TO_PLAY_STEP_BACKWARD,
});
export const howToPlayStateForward = () => ({
  type: HOW_TO_PLAY_STATE_FORWARD,
});
