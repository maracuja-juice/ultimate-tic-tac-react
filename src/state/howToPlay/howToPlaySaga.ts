import { delay, put, takeEvery, race, call, take } from 'redux-saga/effects';
import {
  HOW_TO_PLAY_CLOSE,
  HOW_TO_PLAY_STEP_BACKWARD,
  HOW_TO_PLAY_STEP_FORWARD,
  howToPlayStateForward,
} from './howToPlayActions';

function* updateBoardState() {
  while (true) {
    yield delay(1150);
    yield put(howToPlayStateForward());
  }
}

function* startUpdate() {
  yield race({
    update: call(updateBoardState),
    cancel1: take(HOW_TO_PLAY_CLOSE),
    cancel2: take(HOW_TO_PLAY_STEP_FORWARD),
    cancel3: take(HOW_TO_PLAY_STEP_BACKWARD),
  });
}

function* howToPlaySaga() {
  yield takeEvery(HOW_TO_PLAY_STEP_FORWARD, startUpdate);
  yield takeEvery(HOW_TO_PLAY_STEP_BACKWARD, startUpdate);
}

export default howToPlaySaga;
