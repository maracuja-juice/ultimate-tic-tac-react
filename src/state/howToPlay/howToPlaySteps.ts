import { TicTacToeGame } from '../../util';
import { Move } from '../AppState';
import { winStates } from './possibleWinStates';

export interface HowToPlayStep {
  text: string;
  states: TicTacToeGame[];
  moves: Move[];
}

// TODO export all texts in the UI into a json file, in the future
export const steps: HowToPlayStep[] = [
  {
    text: 'Ultimate Tic-Tac-Toe is played with 9 small tic-tac-toe boards',
    states: [new TicTacToeGame([])],
    moves: [],
  },
  {
    text: 'Each turn a player marks a field',
    states: [new TicTacToeGame([])],
    moves: [
      { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 0, y: 0 } },
      { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 0, y: 1 } },
    ],
  },
  {
    text:
      'When you get three in a row on a small board, youâ€™ve won that board.',
    states: [
      new TicTacToeGame([
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 1, y: 0 } },
        { boardPosition: { x: 1, y: 0 }, tilePosition: { x: 0, y: 0 } },
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 1, y: 1 } },
      ]),
    ],
    moves: [
      { boardPosition: { x: 1, y: 1 }, tilePosition: { x: 0, y: 0 } },
      { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 1, y: 2 } },
    ],
  },
  {
    text: 'To win the game, you need to win three small boards in a row',
    states: winStates,
    moves: [],
  },
  {
    // TODO insert component to show that this is marked by light blue?
    // TODO: animation when going to step is not turned off!?
    text:
      'Each turn the previous move of your opponent dictates in which board you can move next',
    states: [new TicTacToeGame([])],
    moves: [
      { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 0, y: 1 } },
      { boardPosition: { x: 0, y: 1 }, tilePosition: { x: 0, y: 2 } },
      { boardPosition: { x: 0, y: 2 }, tilePosition: { x: 1, y: 0 } },
      { boardPosition: { x: 1, y: 0 }, tilePosition: { x: 1, y: 1 } },
      { boardPosition: { x: 1, y: 1 }, tilePosition: { x: 1, y: 2 } },
      { boardPosition: { x: 1, y: 2 }, tilePosition: { x: 2, y: 0 } },
      { boardPosition: { x: 2, y: 0 }, tilePosition: { x: 2, y: 1 } },
      { boardPosition: { x: 2, y: 1 }, tilePosition: { x: 2, y: 2 } },
      { boardPosition: { x: 2, y: 2 }, tilePosition: { x: 0, y: 0 } },
    ],
  },
  {
    text: `If your opponent sends you to a board that's already won, you can make your move on any of the other boards`,
    states: [
      new TicTacToeGame([
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 1, y: 0 } },
        { boardPosition: { x: 1, y: 0 }, tilePosition: { x: 0, y: 0 } },
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 1, y: 1 } },
        { boardPosition: { x: 1, y: 1 }, tilePosition: { x: 0, y: 0 } },
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 1, y: 2 } },
      ]),
    ],
    moves: [
      { boardPosition: { x: 1, y: 2 }, tilePosition: { x: 2, y: 2 } },
      { boardPosition: { x: 2, y: 2 }, tilePosition: { x: 0, y: 0 } },
      { boardPosition: { x: 2, y: 0 }, tilePosition: { x: 1, y: 2 } },
    ],
  },
  {
    text:
      'If one of the small boards results in a tie, the board counts for neither X nor O.',
    states: [
      new TicTacToeGame([
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 2, y: 0 } },
        { boardPosition: { x: 2, y: 0 }, tilePosition: { x: 0, y: 0 } },
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 0, y: 2 } },
        { boardPosition: { x: 0, y: 2 }, tilePosition: { x: 1, y: 2 } },
        { boardPosition: { x: 1, y: 2 }, tilePosition: { x: 0, y: 0 } },
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 1, y: 1 } },
        { boardPosition: { x: 1, y: 1 }, tilePosition: { x: 0, y: 0 } },
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 0, y: 0 } },
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 2, y: 2 } },
        { boardPosition: { x: 2, y: 2 }, tilePosition: { x: 2, y: 2 } },
        { boardPosition: { x: 2, y: 2 }, tilePosition: { x: 0, y: 0 } },
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 2, y: 1 } },
        { boardPosition: { x: 2, y: 1 }, tilePosition: { x: 2, y: 1 } },
        { boardPosition: { x: 2, y: 1 }, tilePosition: { x: 0, y: 0 } },
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 0, y: 1 } },
        { boardPosition: { x: 0, y: 1 }, tilePosition: { x: 0, y: 1 } },
        { boardPosition: { x: 0, y: 1 }, tilePosition: { x: 0, y: 0 } },
        { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 1, y: 2 } },
        { boardPosition: { x: 1, y: 2 }, tilePosition: { x: 1, y: 0 } },
      ]),
    ],
    moves: [
      { boardPosition: { x: 1, y: 0 }, tilePosition: { x: 0, y: 0 } },
      { boardPosition: { x: 0, y: 0 }, tilePosition: { x: 1, y: 0 } },
    ],
  },
];
