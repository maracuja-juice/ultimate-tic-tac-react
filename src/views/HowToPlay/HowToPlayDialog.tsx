import * as React from 'react';
import {
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
} from '@rmwc/dialog';
import styles from './HowToPlayDialog.module.css';
import { HowToPlayContent } from './HowToPlayContent';
import { HowToPlayBoardState } from '../../state/AppState';
import { DialogActionBar } from './DialogActionBar';
import { useKeyPressEvent } from 'react-use';
import { Keys } from '../../lib';

export interface HowToPlayDialogProps {
  onClose: () => void;
  onForward: () => void;
  onBackward: () => void;
  stepNumber: number;
  maxStepNumber: number;
  text: React.ReactNode;
  boardState: HowToPlayBoardState;
}

export function HowToPlayDialog(props: HowToPlayDialogProps) {
  const {
    onClose,
    onForward,
    onBackward,
    stepNumber,
    maxStepNumber,
    text,
    boardState,
  } = props;

  // preventOutsideDismiss also disables Escape key, so we enable it manually
  useKeyPressEvent(Keys.Escape, onClose, null);

  return (
    <Dialog
      className={styles.dialog}
      open={true}
      onClose={onClose}
      preventOutsideDismiss={true}
    >
      <DialogTitle>How to play</DialogTitle>
      <DialogContent>
        <p>{text}</p>
        <div className={styles.bigBoard}>
          <HowToPlayContent boardState={boardState} />
        </div>
      </DialogContent>
      <DialogActions>
        <DialogActionBar
          onClose={onClose}
          onForward={onForward}
          onBackward={onBackward}
          stepNumber={stepNumber}
          maxStepNumber={maxStepNumber}
        />
      </DialogActions>
    </Dialog>
  );
}
