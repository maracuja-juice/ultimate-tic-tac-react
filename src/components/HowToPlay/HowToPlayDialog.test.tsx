import * as React from 'react';
import { shallow } from 'enzyme';
import { HowToPlayDialog } from './HowToPlayDialog';
import { howToPlaySteps } from '../../state/howToPlay/howToPlaySteps';

describe('HowToPlayDialog', () => {
  it('should match snapshot', () => {
    const step = howToPlaySteps[1];
    const stepState = step.states[0];
    const howToPlayDialog = shallow(
      <HowToPlayDialog
        boardState={{
          board: stepState.getBoard(),
          currentPlayer: stepState.getCurrentPlayer(),
          animate: true,
          activeBoards: stepState.getCurrentActiveBoards(),
        }}
        onOpen={() => {}}
        onClose={() => {}}
        text={step.text}
        onForward={() => {}}
        onBackward={() => {}}
        stepNumber={1}
        maxStepNumber={10}
      />,
    );

    expect(howToPlayDialog).toMatchSnapshot();
  });
});
