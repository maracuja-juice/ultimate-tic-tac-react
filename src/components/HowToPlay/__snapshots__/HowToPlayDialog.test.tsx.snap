// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HowToPlayDialog should match snapshot 1`] = `
<Function
  className="dialog"
  onClose={[Function]}
  open={true}
  preventOutsideDismiss={true}
>
  <Function>
    How to play
  </Function>
  <Function>
    <p>
      Each turn a player marks a field
    </p>
    <div
      className="bigBoard"
    >
      <HowToPlayContent
        boardState={
          Object {
            "activeBoards": Array [
              Object {
                "x": 0,
                "y": 0,
              },
              Object {
                "x": 0,
                "y": 1,
              },
              Object {
                "x": 0,
                "y": 2,
              },
              Object {
                "x": 1,
                "y": 0,
              },
              Object {
                "x": 1,
                "y": 1,
              },
              Object {
                "x": 1,
                "y": 2,
              },
              Object {
                "x": 2,
                "y": 0,
              },
              Object {
                "x": 2,
                "y": 1,
              },
              Object {
                "x": 2,
                "y": 2,
              },
            ],
            "animate": true,
            "board": Array [
              Object {
                "position": Object {
                  "x": 0,
                  "y": 0,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 0,
                  "y": 1,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 0,
                  "y": 2,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 1,
                  "y": 0,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 1,
                  "y": 1,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 1,
                  "y": 2,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 2,
                  "y": 0,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 2,
                  "y": 1,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
              Object {
                "position": Object {
                  "x": 2,
                  "y": 2,
                },
                "tiles": Array [
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 0,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 1,
                      "y": 2,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 0,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 1,
                    },
                    "value": 2,
                  },
                  Object {
                    "boardPosition": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "position": Object {
                      "x": 2,
                      "y": 2,
                    },
                    "value": 2,
                  },
                ],
                "value": 2,
              },
            ],
            "currentPlayer": 0,
          }
        }
      />
    </div>
  </Function>
  <Function>
    <ActionBar
      maxStepNumber={10}
      onBackward={[Function]}
      onClose={[Function]}
      onForward={[Function]}
      stepNumber={1}
    />
  </Function>
</Function>
`;
